packages/data/src/types.ts

```diff
@@ -37,6 +37,41 @@ export interface ReduxStoreConfig<
 	controls?: MapOf< Function >;
 }
 
+export type UseSelectReturn<
+	F extends MapSelect | StoreDescriptor< any >
+> = F extends MapSelect
+	? ReturnType< F >
+	: F extends StoreDescriptor< any >
+	? CurriedSelectorsOf< F >
+	: never;
+
+export type MapSelect = ( select: SelectFunction ) => any;
+
+export type SelectFunction = < S >( store: S ) => CurriedSelectorsOf< S >;
+
+export type CurriedSelectorsOf< S > = S extends StoreDescriptor<
+	ReduxStoreConfig< any, any, infer Selectors >
+>
+	? { [ key in keyof Selectors ]: CurriedState< Selectors[ key ] > }
+	: never;
+
+/**
+ * Removes the first argument from a function
+ *
+ * This is designed to remove the `state` parameter from
+ * registered selectors since that argument is supplied
+ * by the editor when calling `select(…)`.
+ *
+ * For functions with no arguments, which some selectors
+ * are free to define, returns the original function.
+ */
+export type CurriedState< F > = F extends (
+	state: any,
+	...args: infer P
+) => infer R
+	? ( ...args: P ) => R
+	: F;
```

Comment:
what was the issue with this inference? is it that we lose the generic type parameters through this inference? so if we had `const get<T> = (state, id): T => state[id];` that it widens to `any`?

Reply:
More or less:

```
type BadlyInferredSignature = CurriedState<
    <K extends string | number>(
        state: any,
        kind: K,
        key: K extends string ? 'one value' : false
    ) => K
>
// BadlyInferredSignature evaluates to:
// (kind: string number, key: false | "one value") => string number
```

The [only solution I found is quite convoluted](https://github.com/WordPress/gutenberg/pull/41578) so I went for shipping the simple Curry type in this PR. We can always ship updated version in a follow-up PR.